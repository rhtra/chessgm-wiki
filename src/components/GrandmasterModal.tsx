import { countryCodeToFlag } from "../utils/flags";
import { useClock } from "../utils/useClock";

type Player = {
  username: string;
  name?: string;
  avatar?: string;
  countryName?: string;
  countryCode?: string;
  fide?: number;
  last_online?: number;
  joined?: number;
  status?: string;
  is_streamer?: boolean;
  verified?: boolean;
  league?: string;
  streaming_platforms?: string[];
};

type Props = {
  player: Player;
  onClose: () => void;
};

export default function GrandmasterModal({ player, onClose }: Props) {
  const elapsed = useClock(player.last_online);

  return (
    <div className="modal-backdrop" onClick={onClose}>
      <div className="modal" onClick={(e) => e.stopPropagation()}>
        <button className="close-btn" onClick={onClose}>
          ‚úï
        </button>

        <img
          className="modal-avatar"
          src={
            player.avatar ||
            `https://ui-avatars.com/api/?name=${encodeURIComponent(
              player.username
            )}&background=000&color=fff`
          }
          alt={player.username}
        />

        <h2>
          GM{" "}
          {player.name ? (
            <>
              {player.name} <span className="username">({player.username})</span>
            </>
          ) : (
            player.username
          )}
        </h2>


        <p>
          Country:{" "}
          {player.countryName ? (
            <>
              {countryCodeToFlag(player.countryCode || "") || "üåç"}{" "}
              {player.countryName}
            </>
          ) : (
            "Unknown"
          )}
        </p>

        <p>Status: {player.status || "Unknown"}</p>
        <p>Verified: {player.verified ? "‚úÖ Yes" : "‚ùå No"}</p>
        <p>Streamer: {player.is_streamer ? "üé• Yes" : "No"}</p>
        <p>League: {player.league || "N/A"}</p>
        {player.streaming_platforms && player.streaming_platforms.length > 0 && (
          <p>
            Platforms: {player.streaming_platforms.join(", ")}
          </p>
        )}

        <p>
          Joined:{" "}
          {player.joined
            ? new Date(player.joined * 1000).toLocaleDateString()
            : "Unknown"}
        </p>

        <p>
          Last Online:{" "}
          {player.last_online
            ? new Date(player.last_online * 1000).toLocaleString()
            : "Unknown"}
        </p>

        {player.last_online && (
          <p>Time since last online: {elapsed}</p>
        )}

        <p>Rating: {player.fide || "No rating"}</p>

        <a
          href={`https://www.chess.com/member/${player.username}`}
          target="_blank"
          rel="noreferrer"
          className="btn"
        >
          Open Chess.com profile
        </a>
      </div>
    </div>
  );
}
