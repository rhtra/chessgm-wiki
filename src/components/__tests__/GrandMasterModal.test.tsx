import { render, screen, fireEvent } from "@testing-library/react";
import GrandmasterModal from "../GrandmasterModal";

// ✅ Mock the named export useClock
jest.mock("../../utils/useClock", () => ({
  __esModule: true,
  useClock: () => "01:00:00",
}));

const player = {
  username: "testgm",
  name: "Test GM",
  countryName: "Norway",
  countryCode: "NO",
  fide: 2800,
  last_online: Math.floor(Date.now() / 1000) - 3600,
  joined: 1600000000,
  status: "basic",
  is_streamer: false,
  verified: true,
  league: "champions",
  streaming_platforms: [],
};

describe("GrandmasterModal snapshots", () => {
  beforeAll(() => {
    jest.spyOn(Date.prototype, "toLocaleString").mockReturnValue("9/30/2025, 8:53:00 PM");
  });

  afterAll(() => {
    (Date.prototype.toLocaleString as jest.Mock).mockRestore();
  });

  it("matches snapshot with player info", () => {
    const { asFragment } = render(
      <GrandmasterModal player={player} onClose={() => {}} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders player details", () => {
    render(<GrandmasterModal player={player} onClose={() => {}} />);

    expect(screen.getByText(/Test GM/i)).toBeInTheDocument();
    expect(screen.getByText(/\(testgm\)/i)).toBeInTheDocument();
    expect(screen.getByText(/Norway/)).toBeInTheDocument();
    expect(screen.getByText(/Rating: 2800/i)).toBeInTheDocument();
    expect(screen.getByText(/01:00:00/i)).toBeInTheDocument(); // ✅ mocked clock
    expect(screen.getByText(/9\/30\/2025, 8:53:00 PM/i)).toBeInTheDocument(); // ✅ stable date
  });

  it("closes when clicking the close button", () => {
    const onClose = jest.fn();
    render(<GrandmasterModal player={player} onClose={onClose} />);
    fireEvent.click(screen.getByText("✕"));
    expect(onClose).toHaveBeenCalled();
  });
});
