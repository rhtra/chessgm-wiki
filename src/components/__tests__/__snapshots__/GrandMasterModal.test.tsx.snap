// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`GrandmasterModal snapshots matches snapshot with player info 1`] = `
<DocumentFragment>
  <div
    class="modal-backdrop"
  >
    <div
      class="modal"
    >
      <button
        class="close-btn"
      >
        âœ•
      </button>
      <img
        alt="testgm"
        class="modal-avatar"
        src="https://ui-avatars.com/api/?name=testgm&background=000&color=fff"
      />
      <h2>
        GM Test GM 
        <span
          class="username"
        >
          (testgm)
        </span>
      </h2>
      <p>
        Country: ðŸ‡³ðŸ‡´ Norway
      </p>
      <p>
        Status: basic
      </p>
      <p>
        Verified: âœ… Yes
      </p>
      <p>
        Streamer: No
      </p>
      <p>
        League: champions
      </p>
      <p>
        Joined: 9/13/2020
      </p>
      <p>
        Last Online: 9/30/2025, 8:53:00 PM
      </p>
      <p>
        Time since last online: 01:00:00
      </p>
      <p>
        Rating: 2800
      </p>
      <a
        class="btn"
        href="https://www.chess.com/member/testgm"
        rel="noreferrer"
        target="_blank"
      >
        Open Chess.com profile
      </a>
    </div>
  </div>
</DocumentFragment>
`;
